prow/git/git.go

```diff
@@ -146,7 +146,8 @@ func (c *Client) unlockRepo(repo string) {
 // In that case, it must do a full git mirror clone. For large repos, this can
 // take a while. Once that is done, it will do a git fetch instead of a clone,
 // which will usually take at most a few seconds.
-func (c *Client) Clone(repo string) (*Repo, error) {
+func (c *Client) Clone(organization, repository string) (*Repo, error) {
```

Comment:
This effectively makes the gitclient unsable for anything that doesn't follow GitHubs org/repo pattern, e.G. `gerrit`. Is that a tradeoff we want to make?

cc @krzyzacy

Reply:
We need to be able to support other git providers, especially gerrit since we already support that today.  If this is important for the GitHub use case we should expose a GitHub specific wrapper around the more generic `Clone` that handles the GitHub specific concatenation.

Reply:
Do we have any code that uses this library for Gerrit today? Are there any tests that verify it? Do we know what support we need in the future?

Reply:
Talked with @alvaroaleman on Slack ... I am not sure I follow the reasoning here -- if we do not use this library today for `gerrit`, why should we be using it in the future? If we do not have concrete use cases we want to support, I don't really want to code in the dark for hypothetical integrations. Once the v2 `git` library is in, everything about its implementation is behind an interface and can be refactored with more ease. Perhaps a different client factory would be used for other types of clones.

Reply:
How about a URL instead of repo and/or org/repo? That's what git actually uses anyway.

Reply:
No, that's really not the point of the client, either ... Do we have a specific business function we are trying to solve here? I am happy to write code to that end but don't necessarily want to write hypotheticals ... disentangling *this* operation from the way that the base is constructed is exactly what enterprise support needs ...

Reply:
The git client is currently provider agnostic and I think we want Prow to be able to work with git providers besides GitHub someday (that is part of what scallywag is all about). I don't see why we would decide to make the git client GitHub specific now. A GitHub specific client factory seems fine, but I don't think the underlying client should be specific to GitHub.
This client might not be used for Gerrit now, but we do use `clonerefs` with Gerrit and we will want `clonerefs` to use the new client, right?

Reply:
It is not. It uses GitHub-specific `pull/N/head` refspecs that are synthetic and only exist on GitHub.

Reply:
FWIW it is also not "GitHub" specific to ask for an org and a repo, that is also how other providers like GitLab work. In any case I just don't understand this bikeshed. This client is useless for gerrit. Nothing we do today requires it to support gerrit. Why is it required for me making a change here to attempt to anticipate heretofore unknown, undocumented and inscrutable requirements? Why don't we ask the developers implementing new features in the future to do that work?

Reply:
Also no, I do not expect `clonerefs` to use this client. We talked this through when `clonerefs` was written. This is a GitHub `git` client with caching that makes plugins work efficiently. `clonerefs` is a library to allow for cloning Prow `refs` and `extra_refs` in a way that allows for a previously-cloned cache to exist. They both use `git` underneath but are separate libraries achieving separate goals and I do not believe there is any utility to combining them. It will not be DRY to do so as we are not repeating ourselves.

Reply:
> Why is it required for me making a change here to attempt to anticipate heretofore unknown, undocumented and inscrutable requirements?

I assumed we wanted to respect the general git interface and support any provider instead of making assumptions that the provider is GitHub (or maybe some similar provider like GitLab), especially given the recent interest in making Prow support other git providers. I wasn't aware that the current client is so tightly coupled to GitHub and I'm certainly not trying to require you to implement unknown or undocumented requirements for the future, I was just thinking that the goal was to write a client to interface with git generally. Sorry for the misunderstanding.
I was also under the impression that we were trying to write a client that could be used for all of our git use cases throughout Prow.

If this client is only for the GitHub (or similar providers) use cases and we make that clear I'm ok with that. We can refactor it again if and when we need to add support for other providers.

Reply:
I'm sorry, I did not make my intent clear here. Yes, this is simply an effort to get the current GitHub-based `git` client used in plugins to be a) more encapsulated and b) have more fields dynamically computed. The intent of these changes is to not change any functionality but just to make the code more sustainable and allow for more modifications in a testable manner. This lays the groundwork for other refactors or additions of utility if they come in the future. I do not necessarily think that I've seen a compelling argument for using one unified library for all interaction with `git` in Prow, so I have not prioritized that.